server/src/main/scala/org/http4s/server/middleware/RequestId.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s
+package server
+package middleware
+
+import org.http4s.{Header, Http, Request, Response}
+import cats.{FlatMap, ~>}
+import cats.arrow.FunctionK
+import cats.data.{Kleisli, OptionT}
+import cats.effect.Sync
+import cats.implicits._
+import org.typelevel.ci.CIString
+import java.util.UUID
+
+/** Propagate a `X-Request-Id` header to the response, generate a UUID
+  * when the `X-Request-Id` header is unset.
+  * https://devcenter.heroku.com/articles/http-request-id
+  */
+object RequestId {
+
+  private val requestIdHeader = CIString("X-Request-ID")
+
+  def apply[G[_], F[_]](
+      fk: F ~> G,
+      headerName: CIString = requestIdHeader,
+      genReqId: Option[F[UUID]] = None
```

Comment:
I wonder about two constructors here instead of this optional parameter, because if we provide some generator, I don't think we need the `Sync` constraint.

Reply:
Okidoke, I'll mess around with the method types and see if I can get something nice working.